File: apps/ai-draw/backend-go/generation_service.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
)

const maxBatchCount = 20

type generationService struct {
	registry    *taskRegistry
	outputStore *imageStore
	provider    imageProvider
}

func newGenerationService(outputStore *imageStore, provider imageProvider, sink eventSink) *generationService {
	if provider == nil {
		provider = newOpenAIImageProvider(nil)
	}
	return &generationService{
		registry:    newTaskRegistry(sink),
		outputStore: outputStore,
		provider:    provider,
	}
}

func (svc *generationService) createNormal(params json.RawMessage) (any, error) {
	var payload struct {
		Request createNormalGenerationRequest `json:"request"`
	}
	if err := decodeRequest(params, &payload); err != nil {
		return nil, err
	}
	if err := validateNormalRequest(payload.Request); err != nil {
		return nil, err
	}
	batchCount := clampInt(payload.Request.BatchCount, 1, maxBatchCount, 1)
	tasks := make([]generationTask, 0, batchCount)
	for i := 0; i < batchCount; i++ {
		req := payload.Request
		req.BatchCount = 1
		ctx, cancel := context.WithCancel(context.Background())
		task := svc.registry.create(generationTaskInput{
			Mode:         generationModeNormal,
			Prompt:       strings.TrimSpace(req.Prompt),
			ProviderID:   strings.TrimSpace(req.Provider.ID),
			ProviderName: strings.TrimSpace(req.Provider.Name),
			Model:        resolveProviderModel(req.Provider),
			Cancel:       cancel,
		})
		tasks = append(tasks, task)
		go svc.runTask(ctx, task.ID, imageGenerationInput{TaskID: task.ID, Mode: generationModeNormal, Normal: &req})
	}
	return map[string]any{"tasks": tasks}, nil
}

func (svc *generationService) createLocalEdit(params json.RawMessage) (any, error) {
	var payload struct {
		Request createLocalEditGenerationRequest `json:"request"`
	}
	if err := decodeRequest(params, &payload); err != nil {
		return nil, err
	}
	if err := validateLocalEditRequest(payload.Request); err != nil {
		return nil, err
	}
	ctx, cancel := context.WithCancel(context.Background())
	req := payload.Request
	task := svc.registry.create(generationTaskInput{
		Mode:         generationModeLocalEdit,
		Prompt:       strings.TrimSpace(req.Prompt),
		ProviderID:   strings.TrimSpace(req.Provider.ID),
		ProviderName: strings.TrimSpace(req.Provider.Name),
		Model:        resolveProviderModel(req.Provider),
		Cancel:       cancel,
	})
	go svc.runTask(ctx, task.ID, imageGenerationInput{TaskID: task.ID, Mode: generationModeLocalEdit, Local: &req})
	return map[string]any{"task": task}, nil
}

func (svc *generationService) get(params json.RawMessage) (any, error) {
	payload, err := decodeMap(params)
	if err != nil {
		return nil, err
	}
	return map[string]any{"task": svc.registry.get(asString(payload["taskId"]))}, nil
}

func (svc *generationService) list(params json.RawMessage) (any, error) {
	payload, err := decodeMap(params)
	if err != nil {
		return nil, err
	}
	return map[string]any{"tasks": svc.registry.list(intNumber(payload["limit"]))}, nil
}

func (svc *generationService) cancel(params json.RawMessage) (any, error) {
	payload, err := decodeMap(params)
	if err != nil {
		return nil, err
	}
	if !svc.registry.cancel(asString(payload["taskId"])) {
		return nil, newDirectError(errorTaskNotFound, "任务不存在")
	}
	return nil, nil
}

func (svc *generationService) dispose() {
	svc.registry.dispose()
}

func (svc *generationService) runTask(ctx context.Context, taskID string, input imageGenerationInput) {
	svc.registry.update(taskID, generationTaskPatch{Status: stringPtr(generationStatusRunning)})
	result, err := svc.provider.Generate(ctx, input)
	if err != nil {
		status := generationStatusFailed
		errorText := err.Error()
		if errors.Is(ctx.Err(), context.Canceled) {
			status = generationStatusCanceled
			errorText = "已取消"
		}
		patch := generationTaskPatch{Status: &status, Error: &errorText}
		var providerErr providerError
		if errors.As(err, &providerErr) && providerErr.Debug != nil {
			patch.Debug = providerErr.Debug
		}
		svc.registry.update(taskID, patch)
		return
	}
	if errors.Is(ctx.Err(), context.Canceled) {
		status := generationStatusCanceled
		errorText := "已取消"
		svc.registry.update(taskID, generationTaskPatch{Status: &status, Error: &errorText, Debug: result.Debug})
		return
	}

	imageDataURL := result.ImageDataURL
	savedPath := ""
	if input.autoSave() {
		path, err := svc.outputStore.saveBase64(result.ImageDataURL)
		if err != nil {
			status := generationStatusFailed
			errorText := err.Error()
			svc.registry.update(taskID, generationTaskPatch{Status: &status, Error: &errorText, Debug: result.Debug})
			return
		}
		savedPath = path
		imageDataURL = ""
	}
	status := generationStatusSucceeded
	svc.registry.update(taskID, generationTaskPatch{Status: &status, ImageDataURL: &imageDataURL, SavedPath: &savedPath, Debug: result.Debug})
}

func decodeRequest(raw json.RawMessage, target any) error {
	if len(raw) == 0 || string(raw) == "null" {
		return newDirectError(errorBadRequest, "生成请求参数为空")
	}
	if err := json.Unmarshal(raw, target); err != nil {
		return newDirectError(errorBadRequest, "生成请求参数无效")
	}
	return nil
}

func validateNormalRequest(req createNormalGenerationRequest) error {
	if strings.TrimSpace(req.Prompt) == "" {
		return newDirectError(errorBadRequest, "提示词为空")
	}
	return validateProviderBasics(req.Provider)
}

func validateLocalEditRequest(req createLocalEditGenerationRequest) error {
	if strings.TrimSpace(req.Prompt) == "" {
		return newDirectError(errorBadRequest, "提示词为空")
	}
	if strings.TrimSpace(req.CropImage.DataURL) == "" {
		return newDirectError(errorBadRequest, "选区图片为空")
	}
	return validateProviderBasics(req.Provider)
}

func validateProviderBasics(provider generationProvider) error {
	if strings.TrimSpace(provider.BaseURL) == "" {
		return newDirectError(errorBadRequest, "Base URL 为空")
	}
	if strings.TrimSpace(provider.APIKey) == "" {
		return newDirectError(errorBadRequest, "API Key 为空")
	}
	if strings.TrimSpace(resolveProviderModel(provider)) == "" {
		return newDirectError(errorBadRequest, "模型为空")
	}
	return nil
}

func clampInt(value int, min int, max int, fallback int) int {
	if value == 0 {
		value = fallback
	}
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

File: apps/ai-draw/src/backend/generation/generationService.ts
import { AI_DRAW_ERROR_CODE, AiDrawDirectError } from '../../shared/errors'
import type { AiDrawCreateLocalEditGenerationRequest, AiDrawCreateNormalGenerationRequest } from '../../shared/domain'
import type { ImageStore } from '../images/imageStore'
import { requestOpenAiImage } from './openAiImageProvider'
import { createTaskRegistry, type TaskRegistry } from './taskRegistry'

export type GenerationService = {
  registry: TaskRegistry
  createNormal(params: unknown): Promise<{ tasks: ReturnType<TaskRegistry['list']> }>
  createLocalEdit(params: unknown): Promise<{ task: NonNullable<ReturnType<TaskRegistry['get']>> }>
  get(params: unknown): Promise<{ task: ReturnType<TaskRegistry['get']> }>
  list(params: unknown): Promise<{ tasks: ReturnType<TaskRegistry['list']> }>
  cancel(params: unknown): Promise<null>
  dispose(): void
}

function providerSnapshot(req: AiDrawCreateNormalGenerationRequest | AiDrawCreateLocalEditGenerationRequest) {
  const provider = req.provider
  const model = String(provider?.customModel || provider?.model || provider?.models?.[0] || '').trim()
  return { providerId: String(provider?.id || '').trim(), providerName: String(provider?.name || '').trim(), model }
}

export function createGenerationService(outputStore: ImageStore): GenerationService {
  const registry = createTaskRegistry()

  async function runTask(taskId: string, request: AiDrawCreateNormalGenerationRequest | AiDrawCreateLocalEditGenerationRequest) {
    const task = registry.get(taskId)
    if (!task) return
    registry.update(taskId, { status: 'running' })
    try {
      const result = await requestOpenAiImage({ taskId, mode: task.mode, request, signal: task.abortController.signal })
      let savedPath = ''
      if (request.autoSave) savedPath = await outputStore.saveBase64(result.imageDataUrl)
      registry.update(taskId, {
        status: 'succeeded',
        imageDataUrl: request.autoSave ? '' : result.imageDataUrl,
        savedPath,
        debug: result.debug,
      })
    } catch (error: any) {
      const aborted = task.abortController.signal.aborted
      registry.update(taskId, {
        status: aborted ? 'canceled' : 'failed',
        error: aborted ? '已取消' : String(error?.message || error || '生成失败'),
        debug: error?.debug || task.debug || null,
      })
    }
  }

  return {
    registry,
    async createNormal(params) {
      const request = (params as any)?.request as AiDrawCreateNormalGenerationRequest
      if (!request || typeof request !== 'object') throw new AiDrawDirectError(AI_DRAW_ERROR_CODE.badRequest, 'generation.createNormal 参数无效')
      const batchCount = Math.max(1, Math.min(20, Math.floor(Number(request.batchCount) || 1)))
      const snap = providerSnapshot(request)
      const tasks = Array.from({ length: batchCount }, () => registry.create({ mode: 'normal', prompt: String(request.prompt || ''), ...snap }))
      for (const task of tasks) void runTask(task.id, { ...request, batchCount: 1 })
      return { tasks }
    },
    async createLocalEdit(params) {
      const request = (params as any)?.request as AiDrawCreateLocalEditGenerationRequest
      if (!request || typeof request !== 'object') throw new AiDrawDirectError(AI_DRAW_ERROR_CODE.badRequest, 'generation.createLocalEdit 参数无效')
      const task = registry.create({ mode: 'local-edit', prompt: String(request.prompt || ''), ...providerSnapshot(request) })
      void runTask(task.id, request)
      return { task }
    },
    async get(params) {
      return { task: registry.get(String((params as any)?.taskId || '')) }
    },
    async list(params) {
      return { tasks: registry.list((params as any)?.limit) }
    },
    async cancel(params) {
      const ok = registry.cancel(String((params as any)?.taskId || ''))
      if (!ok) throw new AiDrawDirectError(AI_DRAW_ERROR_CODE.taskNotFound, '任务不存在')
      return null
    },
    dispose() {
      registry.dispose()
    },
  }
}
